backtester/funding/item.go

```
@@ -0,0 +1,166 @@
+package funding
+
+import (
+	"fmt"
+
+	"github.com/shopspring/decimal"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+)
+
+// Reserve allocates an amount of funds to be used at a later time
+// it prevents multiple events from claiming the same resource
+func (i *Item) Reserve(amount decimal.Decimal) error {
+	if amount.LessThanOrEqual(decimal.Zero) {
+		return errZeroAmountReceived
+	}
+	if amount.GreaterThan(i.available) {
+		return fmt.Errorf("%w for %v %v %v. Requested %v Available: %v",
+			errCannotAllocate,
+			i.exchange,
+			i.asset,
+			i.currency,
+			amount,
+			i.available)
+	}
+	i.available = i.available.Sub(amount)
+	i.reserved = i.reserved.Add(amount)
+	return nil
+}
+
+// Release reduces the amount of funding reserved and adds any difference
+// back to the available amount
+func (i *Item) Release(amount, diff decimal.Decimal) error {
+	if amount.LessThanOrEqual(decimal.Zero) {
+		return errZeroAmountReceived
+	}
+	if diff.IsNegative() && !i.asset.IsFutures() {
+		return fmt.Errorf("%w diff", errNegativeAmountReceived)
+	}
+	if amount.GreaterThan(i.reserved) {
+		return fmt.Errorf("%w for %v %v %v. Requested %v Reserved: %v",
+			errCannotAllocate,
+			i.exchange,
+			i.asset,
+			i.currency,
+			amount,
+			i.reserved)
+	}
+	i.reserved = i.reserved.Sub(amount)
+	i.available = i.available.Add(diff)
+	return nil
+}
+
+// IncreaseAvailable adds funding to the available amount
+func (i *Item) IncreaseAvailable(amount decimal.Decimal) {
+	if amount.IsNegative() || amount.IsZero() {
+		return
```

I would return an error. Up to you.